File: js/src/evm/memory.js
import Element from './element.js'

export default class Memory {
  constructor() {
    this.data = []
  }

  toString() {
    let r = `${this.data.length} elems:\n`
    r += this.data.map(([off, val]) => `  - ${off}: ${val.toString()}`).join('\n')
    return r
  }

  store(offset, value) {
    this.data.push([offset, value])
  }

  size() {
    if (this.data.length === 0) {
      return 0
    }
    return Math.max(...this.data.map(([off, val]) => off + val.data.length))
  }

  get(offset) {
    for (let d = this.data.length - 1; d >= 0; d--) {
      const [off, val] = this.data[d]
      if (off == offset) {
        return val
      }
    }
    return undefined
  }

  load(offset) {
    const ret = new Uint8Array(32)
    const used = new Set()

    for (let idx = 0; idx < 32; idx++) {
      const i = idx + offset
      for (let d = this.data.length - 1; d >= 0; d--) {
        const [off, val] = this.data[d]
        if (i >= off && i < off + val.data.length) {
          // early return if it's one full element
          if (val.label !== undefined) {
            used.add(val.label)
          }
          if (idx === 0 && offset === off && val.data.length === 32) {
            return [val, used]
          }
          ret[idx] = val.data[i - off]
          break
        }
      }
    }

    return [new Element(ret), used];
  }
}

File: evmole/evm/memory.py
from .element import Element


class Memory:
    def __init__(self) -> None:
        self.data: list[tuple[int, Element]] = []

    def __str__(self):
        r = f'{len(self.data)} elems:\n'
        return r + '\n'.join(f'  - {off}: {val.data.hex()} | {val.label}' for off, val in self.data)

    def store(self, offset: int, value: Element):
        self.data.append((offset, value))

    def size(self) -> int:
        if len(self.data) == 0:
            return 0
        return max(off + len(val.data) for off, val in self.data)

    def get(self, offset: int):
        for off, val in reversed(self.data):
            if off == offset:
                return val
        return None

    def load(self, offset: int) -> tuple[Element, set]:
        used = set()
        res = [b'\x00'] * 32
        for idx in range(32):
            i = idx + offset
            for off, val in reversed(self.data):
                if i >= off and i < off + len(val.data):
                    if val.label is not None:
                        used.add(val.label)
                    # early return if it's one full element
                    if idx == 0 and offset == off and len(val.data) == 32:
                        return val, used
                    res[idx] = val.data[i - off : i - off + 1]
                    break

        return Element(data=b''.join(res)), used
